models/relation.rb

```diff
@@ -43,16 +43,18 @@ class Relation < Base
 
     def self.create(hash)
       # TODO: remove it after the linked_ids is replaced with linked_service_ids in the api definition
-      exclusive = [:linked_service_ids, :linked_ids, :linked_services].freeze
-      raise "#{exclusive} fields are mutually exclusive" if hash.keys.count{ |k| exclusive.include? k } > 1
+      if hash.key?(:linked_service_ids) && hash.key?(:linked_ids) && hash[:linked_service_ids] != hash[:linked_ids]
+        raise 'linked_service_ids and linked_ids should have the same content if both used'
+      end
+
+      # Raise condition required due to the use of lib/helper deep_copy
+      raise 'linked_services should not belong to vrp hash' if hash.key?(:linked_services)
```

Comment:
Why do we need this? As far as I remember, deep_copy should be able to handle linked_services/linked_service_ids/linked_ids correctly.

Reply:
Note that it is possible to create a vehicle object with a hash which has `start_point_id` (pointing to a real point object with this id`) but also with a hash which has `start_point` (which points to a real point object directly and creator of vehicle will set the `start_point_id` correctly using this proper point object). So this is the same thing with respect to `linked_services` vs `linked_service_ids`, no?

Reply:
The problem araises if you supply both `start_point_id` and `start_point` and they reference to two diffrent objects. I believe in this case one of them overrides the other but I am not sure which one (it might be order dependent) but overall I think `linked_service_ids` and `linked_services` should work exactly the same I think. 

Except that we have this extra `linked_ids` to maintain :face_exhaling: but that shouldn't make us change the behavior of another field of the model.

Reply:
(if we can help it)

Reply:
I tried to comply with the previous logic in place. But for relations, we did not use deep_copy anymore. It could erase this raise if you think it is not useful anymore.

Reply:
If we erase the raise, we will ignore the `linked_services` silently, no? If it is not going to be ignored we can erase it for sure.


Even if do not use deep_copy we can have linked_services and linked_service_ids at the same time, we just need to figure out which one has the higher "priority" -- using start_point vs start_point_id in a vehicle.create() -- and then let linked_ids get overwritten via the correct info.

I feel like the easiest would be just get rid of linked_ids from the model and just do the linked_ids -> linked_service_ids "swap" before the first vrp.create  (at the same time we do shipment -> service swap) and call it a day. What do you think? 

That way we can mark linked_ids depreciated and add mutually_exclusive linked_service_ids (which is more inline with linked_vehicle_ids) so that we can completely remove it a v0.2?

Reply:
This would save us the relation_hash.delete(:linked_ids) logic and we can completely remove this custom logic of relation.create.

Reply:
Get rid of linked_ids is way more ambitious and requires at least its own dedicated PR. As there is the whole shipment logic to adapt : https://github.com/Mapotempo/optimizer-api/blob/062360a7b59a1e30e9e74596c12828434d93d107/test/models/vrp_test.rb#L150

deep_copy of relations has been replaced by .as_json which did not dump the services but only the linked_service_ids. So even if we want to directly assign services, i assume it should be done outside/after the create.

Reply:
I didn't quite get the shipment part and why it is ambitious. We can do the linked_ids -> linked_service_ids replacement after the shipments turned into services but before anything else. 

Is it not as easy as "find_all linked_ids, replace_all with linked_service_ids" inside the code-code and a "ensure_compatibility" kind of logic that does the swap in vrp.rb (and a coercer in tsphelper) plus a mutually_exclusive `linked_service_ids` in the vrp_input.rb? 

We are trying to converge to a solution for these linked_services/linked_service_ids/linked_ids since multiple days, and what I am trying to prevent is having this extra reassignment/delete and custom create logic. These are there all because I didn't replace all linked_ids with linked_service_ids in the first place and they are a clear pain point.

Reply:
My point is that it turns a +58 -20 9 files PR into a +199 -165 18 files for a change which was not mandatory for the point addressed. Anyway, it is now done